perch/main.py

```diff
@@ -0,0 +1,103 @@
+import json
+import re
+import requests
+
+from .endpoints import ENDPOINTS
+
+
+class CrudBase(object):
+    def get(self):
+        raise NotImplementedError
+
+    def create(self):
+        raise NotImplementedError
+
+    def update(self):
+        raise NotImplementedError
+
+    def delete(self):
+        raise NotImplementedError
+
+
+class EndpointCrud(CrudBase):
+    def __init__(self, api, name=None, path=None, verbs=()):
+        self.api = api
+        self.name = name
+        self.path = path
+        for verb in verbs:
+            setattr(self, self.verb_to_name(verb), self.build_request(verb))
+
+    @staticmethod
+    def verb_to_name(verb):
+        verb = verb.lower()
+        name_map = {
+            'post': 'create',
+            'put': 'update'
+        }
+        return name_map.get(verb, verb)
+
+    def build_url(self, verb, **kwargs):
+        path = self.path
+        params = re.findall(':(.[a-z+_]*)/*', path)
+        for param in params:
+            arg = str(kwargs.get(param, ''))
+            if not arg:
+                if not path.endswith(param) or verb in ('put', 'delete',):
+                    raise TypeError('Squwaaaak! The {} kwarg is required!'.format(param))
+            path = path.replace(':' + param, arg)
+
+        url = self.api.root_url + path.rstrip('/')
+        return url
+
+    def make_request(self, req, **kwargs):
+        url = self.build_url(req.__name__, **kwargs)
+        res = req(url, headers=self.api.headers, json=kwargs)
+        if not str(res.status_code).startswith('2'):
+            raise RuntimeError('Squwaaaak! The api returned an error!\n{}\n'.format(res.reason))
+        body = res.json()
```

Comment:
This will fail with an empty response body, e.g. 204s returned by a DELETE endpoint